mip/driver_imx_rt1020.c

```diff
@@ -0,0 +1,302 @@
+static bool mip_driver_imx_rt1020_up(void *);
+
+const uint32_t EIMR_RX_ERR = 0x2400000;              // Intr mask RXF+EBERR
+
+void ETH_IRQHandler(void);
+static bool mip_driver_imx_rt1020_init(uint8_t *, void *);
+void wait_phy_complete(void);
+
+enum { PHY_ADDR = 0x02, PHY_BCR = 0, PHY_BSR = 1 };     // PHY constants
+
+void delay(uint32_t);
+void delay (uint32_t di) {
+  volatile int dno = 0; // Prevent optimization
+  for (uint32_t i = 0; i < di; i++)
+    for (int j=0; j<20; j++) // PLLx20 (500 MHz/24MHz)
+      dno++;
+}
+
+void wait_phy_complete(void);
+void wait_phy_complete(void) {
+  delay(0x00010000);
+  const uint32_t delay_max = 0x00100000;
+  uint32_t delay_cnt = 0;
+  while (!(ENET->EIR & BIT_SET(23)) && (delay_cnt < delay_max))
+  {delay_cnt++;}
+  ENET->EIR |= BIT_SET(23); // MII interrupt clear
+}
+
+static uint32_t eth_read_phy(uint8_t addr, uint8_t reg) {
+  ENET->EIR |= BIT_SET(23); // MII interrupt clear
+  uint32_t mask_phy_adr_reg = 0x1f; // 0b00011111: Ensure we write 5 bits (Phy address & register)
+  uint32_t phy_transaction = 0x00;
+  phy_transaction = (0x1 << 30) \
+                  | (0x2 << 28) \
+                  | ((uint32_t)(addr & mask_phy_adr_reg) << 23) \
+                  | ((uint32_t)(reg & mask_phy_adr_reg)  << 18) \
+                  | (0x2 << 16);
+
+  //MG_INFO(("phy_transaction %x", phy_transaction));
+  ENET->MMFR = phy_transaction;
+  wait_phy_complete();
+
+  // volatile uint32_t phy_read_out = ENET->MMFR;
+  //MG_INFO(("phy_read_out 0x%x", phy_read_out));
+  return (ENET->MMFR & 0x0000ffff);
+}
+
+static void eth_write_phy(uint8_t addr, uint8_t reg, uint32_t val) {
+  ENET->EIR |= BIT_SET(23); // MII interrupt clear
+  uint8_t mask_phy_adr_reg = 0x1f; // 0b00011111: Ensure we write 5 bits (Phy address & register)
+  uint32_t mask_phy_data = 0x0000ffff; // Ensure we write 16 bits (data)
+  addr &= mask_phy_adr_reg;
+  reg &= mask_phy_adr_reg;
+  val &= mask_phy_data;
+  uint32_t phy_transaction = 0x00;
+  phy_transaction = (uint32_t)(0x1 << 30) \
+                  | (uint32_t)(0x1 << 28) \
+                  | (uint32_t)(addr << 23) \
+                  | (uint32_t)(reg  << 18) \
+                  | (uint32_t)(0x2 << 16) \
+                  | (uint32_t)(val);
+  ENET->MMFR = phy_transaction;
+  wait_phy_complete();
+}
+
+// FEC RX/TX descriptors (Enhanced descriptor not enabled)
+// Descriptor buffer structure, little endian
+
+typedef struct enet_bd_struct_def
+{
+    uint16_t length;  // Data length
+    uint16_t control; // Control and status
+    uint32_t *buffer; // Data ptr
+} enet_bd_struct_t;
+
+// Descriptor and buffer globals, in non-cached area, 64 bits aligned.
+
+__attribute__((section("NonCacheable,\"aw\",%nobits @"))) enet_bd_struct_t rx_buffer_descriptor[(ENET_RXBD_NUM)] __attribute__((aligned((64U))));
+__attribute__((section("NonCacheable,\"aw\",%nobits @"))) enet_bd_struct_t tx_buffer_descriptor[(ENET_TXBD_NUM)] __attribute__((aligned((64U))));
```

Comment:
Is this ARM GCC specific ? Hope Keil handles it

Reply:
Yes, both Keil compilers support that dialect.
This is one of the reasons CMSIS exists...

Reply:
Ok thanks Sergio